File: solutions/215.kth-largest-element-in-an-array.java
/*
 * @lc app=leetcode id=215 lang=java
 *
 * [215] Kth Largest Element in an Array
 */

// @lc code=start
class Solution {
    public int findKthLargest(int[] nums, int k) {
        return quickSelect(nums, 0, nums.length-1, k-1);
    }
    
    int quickSelect(int[] nums, int left, int right, int k) {
        int p = left, pivot = nums[right];
        for(int i=left; i<right; ++i) { // i<right 因为 最右边是 pivot（不动）
            if(nums[i]>=pivot) {
                swap(nums, i, p);
                ++p;
            }
        }
        swap(nums, p, right); // 最后把pivot放置到分界 p 就是 pivotIndex
        if(p<k) return quickSelect(nums, p+1, right, k);
        if(p>k) return quickSelect(nums, left, p-1, k);
        return nums[k];
        
    }
    
    void swap(int[] nums, int i, int j) {
        var temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}
// @lc code=end

File: solutions/215.kth-largest-element-in-an-array.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def find_kth_largest(nums, k)
  quick_select(nums, 0, nums.size-1, k-1)
end

def quick_select(nums, s, e, k)
  p = s
  pivot = nums[e]
  (s..e).each do |j|
    if nums[j] >= pivot
      swap(nums, p, j)
      p += 1
    end
  end
  return quick_select(nums, p, e, k) if p-1 < k
  return quick_select(nums, s, p-2, k) if p-1 > k
  nums[p-1]
end

def swap(nums, i, j)
  nums[i], nums[j] = nums[j], nums[i]
end
